src/t_hash.c

```diff
@@ -32,233 +32,119 @@
- * provided defrag functions. The defrag functions return NULL if the allocation
- * was not moved, otherwise they return a pointer to the new memory location.
- * A separate sds defrag function is needed because of the unique memory layout
- * of sds strings.
- * If the location of the hashTypeEntry changed we return the new location,
- * otherwise we return NULL. */
-hashTypeEntry *hashTypeEntryDefrag(hashTypeEntry *entry, void *(*defragfn)(void *), sds (*sdsdefragfn)(sds)) {
-    if (entryHasValuePtr(entry)) {
-        sds *value_ref = hashTypeEntryGetValueRef(entry);
-        sds new_value = sdsdefragfn(*value_ref);
-        if (new_value) *value_ref = new_value;
-    }
-    char *allocation = hashTypeEntryAllocPtr(entry);
-    char *new_allocation = defragfn(allocation);
-    if (new_allocation != NULL) {
-        /* Return the same offset into the new allocation as the entry's offset
-         * in the old allocation. */
-        return new_allocation + ((char *)entry - allocation);
-    }
-    return NULL;
+static void hashTypeDeleteVolatileSet(robj *o) {
+    volatile_set **volatile_set_ref = hashtableMetadata(o->ptr);
+    freeVolatileSet(*volatile_set_ref);
+    *volatile_set_ref = NULL;
+    /* serves mainly for optimization. by changing the hashtable type we can avoid extra function call in hashtable access */
+    hashTypeIgnoreTTL(o, true);
+}
+
+void hashTypeTrackEntry(robj *o, void *entry) {
+    volatile_set *set = hashTypeGetOrcreateVolatileSet(o);
+    serverAssert(volatileSetAddEntry(set, entry, entryGetExpiry(entry)));
 }
 
-/* Used for releasing memory to OS to avoid unnecessary CoW. Called when we've
- * forked and memory won't be used again. See zmadvise_dontneed() */
-void dismissHashTypeEntry(hashTypeEntry *entry) {
-    /* Only dismiss values memory since the field size usually is small. */
-    if (entryHasValuePtr(entry)) {
-        dismissSds(*hashTypeEntryGetValueRef(entry));
+void hashTypeUntrackEntry(robj *o, void *entry) {
+    if (!entryHasExpiry(entry)) return;
+    volatile_set *set = hashTypeGetVolatileSet(o);
+    debugServerAssert(set);
+    serverAssert(volatileSetRemoveEntry(set, entry, entryGetExpiry(entry)));
+    if (volatileSetNumEntries(set) == 0) {
+        hashTypeDeleteVolatileSet(o);
     }
 }
 
-void freeHashTypeEntry(hashTypeEntry *entry) {
-    if (entryHasValuePtr(entry)) {
-        sdsfree(*hashTypeEntryGetValueRef(entry));
+void hashTypeTrackUpdateEntry(robj *o, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry) {
+    int old_tracked = (old_entry && old_expiry != EXPIRY_NONE);
+    int new_tracked = (new_entry && new_expiry != EXPIRY_NONE);
+    /* If entry was not tracked before and not going to be tracked now, we can simply return */
+    if (!old_tracked && !new_tracked)
+        return;
+
+    volatile_set *set = hashTypeGetOrcreateVolatileSet(o);
+    debugServerAssert(set);
+
+    if (old_tracked && !new_tracked)
```

Comment:
nit - use curly braces to match the `else` branch
```suggestion
    if (old_tracked && !new_tracked) {
```